src/main/java/org/jruby/ext/openssl/x509store/StoreContext.java

```diff
@@ -661,6 +665,26 @@ public int verifyCertificate() throws Exception {
         List<X509AuxCertificate> sktmp = null;
         if ( untrusted != null ) {
             sktmp = new ArrayList<X509AuxCertificate>(untrusted);
+
+            // replace certs in untrusted with trusted versions if found
+            X509Object[] objTmp = {null};
+            for (ListIterator<X509AuxCertificate> iter = sktmp.listIterator(); iter.hasNext();) {
+                X509AuxCertificate skCert = iter.next();
+                X500Principal principal = skCert.cert.getSubjectX500Principal();
+                int ok = getBySubject(X509Utils.X509_LU_X509, new Name(principal), objTmp);
+                if (ok == X509Utils.X509_LU_X509) {
+                    // replace old with new and clear rest of untrusted
+                    Certificate certificate = (Certificate) objTmp[0];
+                    if (certificate.x509.equals(skCert)) {
```

Comment:
I am sure I am missing bits and pieces. the question for me is what cert-chain do I need to send to get it trusted. so I need the `skCert` to match the subject-principal of a trusted certificate and then I need to satisfy this equals method.

from the supplied threads I figured that we basically want to replace self-signed certificates in the middle of the chain by some trusted certificate from local trust-store.

Reply:
@mkristian Yes, that's exactly it. I'm working on getting the right things to compare to allow an untrusted cert from the wire to properly match a trusted cert in the local store. Need to read more of the OpenSSL code to understand exactly how much needs to be the same.

Reply:
@headius I was pounding a bit more on this. maybe this is just OK as is when there new trusted certificate chain gets verified after this. after all that is what we want to achieve: a trusted verified chain from the server certificate to root certificate where root comes from our root certificate store.